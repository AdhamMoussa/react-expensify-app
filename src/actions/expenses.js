import uuid from "uuid";

export const addExpense = ({
  description = "",
  amount = 0,
  note = "",
  createdAt = 0
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    newExpense: {
      id: uuid(),
      description,
      amount,
      note,
      createdAt
    }
  };
};

export const editExpense = (id, updatedData) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updatedData
  };
};

export const removeExpense = (id) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  };
};
