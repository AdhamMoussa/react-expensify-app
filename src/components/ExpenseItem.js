import React from "react";
import { Link } from 'react-router-dom';

const ExpenseItem = (props) => {
  return (
    <div>
      <Link to={`/edit/${props.id}`}>
        <h3>Expense: {props.description}</h3>
      </Link>
      <p>Amount: {props.amount}</p>
      <p>Date: {props.createdAt.format('MMM Do, YYYY')}</p>
    </div>
  );
};

export default ExpenseItem;
