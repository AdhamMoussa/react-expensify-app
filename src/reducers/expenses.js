const defaultExpensesState = [];

const expensesReducer = (state = defaultExpensesState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.newExpense];
    case "EDIT_EXPENSE":
      return state.map(
        expense =>
          expense.id === action.id
            ? { ...expense, ...action.updatedData }
            : expense
      );
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
    default:
      return state;
  }
};

export default expensesReducer;
