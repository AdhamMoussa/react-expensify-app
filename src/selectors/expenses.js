import moment from 'moment';

const getExpensesView = (expenses, filters) => {
  return expenses.filter((expense) => {
    const textMatch = expense.description.toLowerCase().includes(filters.text.toLowerCase());
    const startDateMatch = filters.startDate ? expense.createdAt.isSameOrAfter(filters.startDate) : true;
    const endDateMatch = filters.endDate ? expense.createdAt.isSameOrBefore(filters.endDate) : true;

    return textMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
    if (filters.sortBy === 'date') {
      return a.createdAt.isAfter(b.createdAt) ? -1 : 1;
    } else if (filters.sortBy === 'amount') {
      return a.amount > b.amount ? -1 : 1;
    }
  });
};

export default getExpensesView;
