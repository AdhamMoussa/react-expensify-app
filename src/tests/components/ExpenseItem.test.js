import React from 'react'
import { shallow } from 'enzyme';
import ExpenseItem from '../../components/ExpenseItem';
import expenses from '../fixtures/demoExpenses';

test('should render ExpenseItem correctly with an Expense', () => {
  const wrapper = shallow(<ExpenseItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});

