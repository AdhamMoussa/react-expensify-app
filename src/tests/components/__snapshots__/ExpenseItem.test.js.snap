// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseItem correctly with an Expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Expense: 
      gas
    </h3>
  </Link>
  <p>
    Amount: 
    51200
  </p>
  <p>
    Date: 
    Jan 1st, 1970
  </p>
</div>
`;
